### PR TITLE
Transcribe

### PR DESCRIPTION
**Pull request by Cofucan**
Implement Transcription Feature
​
## Description
- I have implemented the transcription feature using Deepgram API.
- I updated the video model to have a column to store the location of the transcript.
- I made a change to the structure in which the recording data is saved. A folder is created for a user, then a folder is created for each video of that user. In that folder will be saved the uploaded video, the transcript and other data like the audio file, the compressed version of the video and the thumbnail
- I have made the video transcription and other processing to be background processes.
- I made some changes to some of the routes.
- I updated the user model and video model to be connected using foreign key relationship.
- Other styling changes
​
## Related Issue (Link to linear ticket)
- A ticket was not created specifically for this, but there is an Issue.
​
## Motivation and Context
- It was requested. It enables a user to see captions when watching their recording.
​
## How Has This Been Tested?
- It was tested using by starting the server and then running a test script on a recorded video.
- After running the script, the appropriate folders and files were created
- Upon checking the database, all records were present and accurate
​
## Screenshots (if appropriate - Postman, etc):
Screenshot from 2023-10-08 16-51-08
Screenshot from 2023-10-08 16-56-42
image

​
## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
​
## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: app/models/user_models.py [hunk 1]
```diff
@@ -1,35 +1,40 @@
...
     __tablename__ = "users"
 
     id: int = Column(Integer, primary_key=True, index=True)
-    username: str = Column(String, index = True, unique=True, nullable=False, default=None)
-    #email: str = Column(String, nullable=True)
-    hashed_password: str = Column(String, unique= True, nullable=False)
+    username: str = Column(
+        String, index=True, unique=True, nullable=False, default=None
+    )
+    # email: str = Column(String, nullable=True)
```
Comment (zxenonx) [inline app/models/user_models.py:16]: let's remove this commented out code

File: app/models/user_models.py [hunk 2]
```diff
@@ -1,35 +1,40 @@
...
+    hashed_password: str = Column(String, unique=True, nullable=False)
     created_date: DateTime = Column(DateTime, server_default=func.now())
     updated_date: DateTime = Column(DateTime, onupdate=func.now())
-    is_deleted: bool = Column(Boolean, default= False)
-    #logged_in: bool = Column(Boolean, default=True)
+    is_deleted: bool = Column(Boolean, default=False)
+    # logged_in: bool = Column(Boolean, default=True)
```
Comment (zxenonx) [inline app/models/user_models.py:21]: this one too

File: app/models/video_models.py [hunk 3]
```diff
@@ -11,26 +11,32 @@ class Video(Base):
...
 
     id: str = Column(String, primary_key=True, unique=True, nullable=False)
-    user_id: str = Column(String, nullable=False)
-    user = relationship("User", back_populates="videos")
+    user_id: int = Column(
+        Integer,
+        ForeignKey("users.id", ondelete="CASCADE"),
+        nullable=False,
+    )
+    # user_id: str = Column(String, nullable=True)
```
Comment (zxenonx) [inline app/models/video_models.py:19]: this one also

File: app/services/services.py [hunk 4]
```diff
@@ -34,33 +38,61 @@ def process_video(
...
+            )
+        )
         compress_video(file_location, compressed_location)
         extract_thumbnail(compressed_location, thumbnail_location)
     except Exception as err:
         # Delete the uploaded file if an error occurs
-        if os.path.exists(file_location):
-            os.remove(file_location)
+        # if os.path.exists(file_location):
+        #     os.remove(file_location)
```
Comment (zxenonx) [inline app/services/services.py:66]: Can we remove those lines?

File: app/services/services.py [hunk 5]
```diff
@@ -34,33 +38,61 @@ def process_video(
...
     video.compressed_location = compressed_location
     video.thumbnail_location = thumbnail_location
-    video.status = "complete"
+    video.status = "completed"
 
     db.commit()
     db.close()
 
 
+def extract_audio(input_path: str, output_path: str) -> None:
```
Comment (zxenonx) [inline app/services/services.py:81]: Doctring missing here

File: app/services/services.py [hunk 6]
```diff
@@ -34,33 +38,61 @@ def process_video(
...
+        "-c:a",
+        "libmp3lame",
+        "-q:a",
+        "2",
+        output_path,
+    ]
+    subprocess.run(command, check=True)
+
+
 def compress_video(input_path: str, output_path: str) -> None:
```
Comment (zxenonx) [inline app/services/services.py:96]: Docstring not complete here

File: app/services/services.py [hunk 7]
```diff
@@ -192,27 +227,54 @@ def merge_blobs(user_id: str, video_id: str) -> str:
...
+    merged_video_path = os.path.join(temp_video_dir, f"{video_id}.mp4")
     with open(merged_video_path, "wb") as merged_file:
         for blob_file in blob_files:
             with open(blob_file, "rb") as f:
                 merged_file.write(f.read())
 
-    # Optionally, remove the temporary directory containing blobs
+    # Optionally, remove the blobs
+    # for blob in blob_files:
+    #     os.remove(blob)
```
Comment (zxenonx) [inline app/services/services.py:238]: We can also remove these lines

File: app/services/services.py [hunk 8]
```diff
@@ -192,27 +227,54 @@ def merge_blobs(user_id: str, video_id: str) -> str:
...
     """
-    
-    ALPHABET = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
-    return str(nanoid.generate(size=10, alphabet=ALPHABET))
+
+    alphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
+    return str(nanoid.generate(size=10, alphabet=alphabet))
+
+
+def get_transcript(audio_file: str, output_path: str) -> None:
```
Comment (zxenonx) [inline app/services/services.py:256]: docstring missing

File: app/services/services.py [hunk 9]
```diff
@@ -192,27 +227,54 @@ def merge_blobs(user_id: str, video_id: str) -> str:
...
+    # Call the async function
+    asyncio.run(generate_transcript(audio_file, output_path, DEEPGRAM_API_KEY))
+
+
+async def generate_transcript(audio_file: str, save_to: str, api_key: str):
```
Comment (zxenonx) [inline app/services/services.py:261]: Docstring missing here as well

File: app/services/services.py [hunk 10]
```diff
@@ -192,27 +227,54 @@ def merge_blobs(user_id: str, video_id: str) -> str:
...
+
+    params = {"punctuate": True, "tier": "enhanced"}
+    with open(audio_file, "rb") as audio:
+        source = {"buffer": audio, "mimetype": "audio/mp3"}
+
+        response = await dg_client.transcription.prerecorded(
+            source, params, timeout=120
+        )
+
+        print(response)
```
Comment (zxenonx) [inline app/services/services.py:272]: We can get rid of this print statement.

File: app/routes/auth_routes.py [hunk 11]
```diff
@@ -1,117 +1,108 @@
...
     user_is_loggedin = is_logged_in(request)
 
     if user_is_loggedin:
         del request.session["username"]
         del request.session["logged_in"]
 
-        return LogoutResponse(status_code=200, message="User Logged out successfully")
+        return LogoutResponse(
+            status_code=200, message="User Logged out successfully"
+        )
```
Comment (zxenonx) [inline app/routes/auth_routes.py:105]: I'd like to suggest we use the guard clause pattern here
`if not user_is_loggedin:
        return LogoutResponse(status_code=401, message="User not logged in")`

  del request.session["username"]
  del request.session["logged_in"]
  return LogoutResponse(
            status_code=200, message="User Logged out successfully"
     )
`

**Comment by DestinedCodes** [comment]
This looks good to me, but when returning information about a video, such as thr get video for a user instead of returning information like compressed, transcript, original and thumbnail location which the Frontend can't make any use of or won't even use them, why not have a Video Response model to rather return url to the video, transcript and thumbnail. I beleive this would be helpful to the front end because when they will make a request to get all videos, they can serialize the response to get the url for all videos and that saves them lot of time than having to get all the IDs and then concatenate them manually btw they wouldn't be using the information of where these files are stored on the server to do anything. What do you think @Cofucan , @zxenonx

**Review by zxenonx — CHANGES_REQUESTED** [review]
Good job @Cofucan 
I left some comments.
Thanks

**Comment by zxenonx** [comment]
> This looks good to me, but when returning information about a video, such as thr get video for a user instead of returning information like compressed, transcript, original and thumbnail location which the Frontend can't make any use of or won't even use them, why not have a Video Response model to rather return url to the video, transcript and thumbnail. I beleive this would be helpful to the front end because when they will make a request to get all videos, they can serialize the response to get the url for all videos and that saves them lot of time than having to get all the IDs and then concatenate them manually btw they wouldn't be using the information of where these files are stored on the server to do anything. What do you think @Cofucan , @zxenonx

I also think having a proper VideoResponse model will help a lot.
@Cofucan will prolly look into it later...
@DestinedCodes

**Comment by Cofucan** [comment]
> > This looks good to me, but when returning information about a video, such as thr get video for a user instead of returning information like compressed, transcript, original and thumbnail location which the Frontend can't make any use of or won't even use them, why not have a Video Response model to rather return url to the video, transcript and thumbnail. I beleive this would be helpful to the front end because when they will make a request to get all videos, they can serialize the response to get the url for all videos and that saves them lot of time than having to get all the IDs and then concatenate them manually btw they wouldn't be using the information of where these files are stored on the server to do anything. What do you think @Cofucan , @zxenonx
> 
> I also think having a proper VideoResponse model will help a lot. @Cofucan will prolly look into it later... @DestinedCodes

Very valid point. I have that in mind actually, I was just trying to make this push for the moment.

**Comment by Cofucan** [comment]
> > This looks good to me, but when returning information about a video, such as thr get video for a user instead of returning information like compressed, transcript, original and thumbnail location which the Frontend can't make any use of or won't even use them, why not have a Video Response model to rather return url to the video, transcript and thumbnail. I beleive this would be helpful to the front end because when they will make a request to get all videos, they can serialize the response to get the url for all videos and that saves them lot of time than having to get all the IDs and then concatenate them manually btw they wouldn't be using the information of where these files are stored on the server to do anything. What do you think @Cofucan , @zxenonx
> 
> I also think having a proper VideoResponse model will help a lot. @Cofucan will prolly look into it later... @DestinedCodes

Done

File: app/routes/auth_routes.py (hunk 11 above)
Comment (Cofucan) [reply app/routes/auth_routes.py:105]: Done

File: app/services/services.py (hunk 10 above)
Comment (Cofucan) [reply app/services/services.py:272]: Done

File: app/services/services.py (hunk 9 above)
Comment (Cofucan) [reply app/services/services.py:261]: Done

File: app/services/services.py (hunk 8 above)
Comment (Cofucan) [reply app/services/services.py:256]: Done

File: app/services/services.py (hunk 5 above)
Comment (Cofucan) [reply app/services/services.py:81]: Done

File: app/models/video_models.py (hunk 3 above)
Comment (Cofucan) [reply app/models/video_models.py:19]: Done

File: app/models/user_models.py (hunk 2 above)
Comment (Cofucan) [reply app/models/user_models.py:21]: Done

File: app/models/user_models.py (hunk 1 above)
Comment (Cofucan) [reply app/models/user_models.py:16]: Done

File: app/services/services.py (hunk 4 above)
Comment (Cofucan) [reply app/services/services.py:66]: Done

File: app/models/user_models.py (hunk 1 above)
Comment (Cofucan) [reply app/models/user_models.py:16]: Done

**Review by zxenonx — APPROVED** [review]
Thanks @Cofucan

**Pull request merged**